### PR TITLE
Add training/inference functions that support multiple input/output arguments and support for custom loss functions via TorchScript.

### PR DESCRIPTION
**Pull request by romerojosh**
This PR introduces more flexible training/inference functions, `torchfort_train_multiarg` and `torchfort_inference_multiarg`, designed to support models that require multiple input/output tensor arguments. To facilitate passing multiple tensors from Fortran/C, we introduce a new type, `torchfort_tensor_list_t`, that are used as arguments to these new routines.

To complete the support for multiple output tensors, we require loss functions that can work with multiple outputs/predictions from the models. The current built-in loss functions expect single prediction/label pairs and are not compatible with multiple model outputs. To enable this, we enable support for custom loss functions exported from TorchScript, similar to our existing support of TorchScript exported models. To maximize flexibility, we include support for custom loss functions with extra tensor arguments (e.g. a mask tensor to ignore data entries in loss computation) via an optional `extra_loss_args` tensor list argument in `torchfort_train_multiarg`.

Please refer to the new Graph Neural Network example added in this PR and new documentation for more details.

This PR also adds new testing for supervised training/inference and loss functions, cover both the newly added and existing functionality.

### CODE REVIEW

File: src/csrc/torchfort.cpp [hunk 1]
```diff
@@ -323,3 +362,73 @@ torchfort_result_t torchfort_load_checkpoint(const char* name, const char* check
...
+
+torchfort_result_t torchfort_tensor_list_create(torchfort_tensor_list_t* tensor_list) {
+  *tensor_list = new torchfort::TensorList;
+  return TORCHFORT_RESULT_SUCCESS;
+}
+
+torchfort_result_t torchfort_tensor_list_destroy(torchfort_tensor_list_t tensor_list_in) {
+  auto tensor_list = static_cast<torchfort::TensorList*>(tensor_list_in);
+  delete tensor_list;
+  tensor_list = nullptr;
```
Comment (azrael417) [inline src/csrc/torchfort.cpp:374]: do we need this?

**Review by azrael417 — APPROVED** [review]
Looks very good, thanks for adding the functionality. I have added some minor comments.

File: CMakeLists.txt [hunk 2]
```diff
@@ -246,11 +248,13 @@ if (TORCHFORT_BUILD_EXAMPLES)
...
     add_subdirectory(examples/fortran/simulation)
+    add_subdirectory(examples/fortran/graph)
   endif()
 endif()
 
 # build tests
 if (TORCHFORT_BUILD_TESTS)
   add_subdirectory(tests/supervised)
   add_subdirectory(tests/rl)
+  add_subdirectory(tests/losses)
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply CMakeLists.txt]: Yep that sounds good. I just renamed the `losses` directory to `general`.

File: examples/fortran/graph/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+This example demostrates the usage of the multi-argument training and inference routines available in TorchFort and how to implement and use a custom loss function.
+
+
+## General Details
+The example code is split one Fortran source file and two PyTorch scripts to generate the model and loss function modules.
+
+[`train.f90`](train.f90) contains the main program and training/inference loops. The simulation is run on an unstructured mesh, defined by the [`nodes.txt`](nodes.txt) and [`connectivity.txt`](connectivity.txt) files.
+The first two columns of the `nodes.txt` file contain the $x$ and $y$ coordinates of the mesh nodes respectively, and the last column contains a node type value, indicating whether the node is on a boundary (`0`) 
+or in the interior (`1`). The `connectivity.txt` file contains a list of triplets, defining the triangles within the unstructured mesh.
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply examples/fortran/graph/README.md]: Done.

File: examples/fortran/graph/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+[`generate_model.py`](generate_model.py) contains a PyTorch script to define a MeshGraphNet-like model. Unlike the `simulation` example, this model take multiple input arguments (`edge_idx`, `edge_feats`, and `node_feats`)
+which requires the use of the `torchfort_train_multiarg` and `torchfort_inference_multiarg` functions to operate. The `edge_idx` input is a list of node index pairs defining the bidirectional edges of the mesh. The `edge_feats` input
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply examples/fortran/graph/README.md]: Done.

File: src/csrc/torchfort.cpp (hunk 1 above)
Comment (romerojosh) [reply src/csrc/torchfort.cpp]: No we don't, especially since the list is passed in my copy anyway. Removed it.

File: docs/usage.rst [hunk 5]
```diff
@@ -124,6 +124,73 @@ For inference, only the model and not a full checkpoint needs to be loaded. For
...
+    istat = torchfort_tensor_list_add_tensor(inputs, input2)
+
+  .. code-tab:: c++
+
+    istat = torchfort_tensor_list_create(inputs);
+    istat = torchfort_tensor_list_add_tensor(inputs, input1, input1_dim, input1_shape, dtype)
+    istat = torchfort_tensor_list_add_tensor(inputs, input2, input2_dim, input2_shape, dtype)
+
+In the Fortran API, ``input1`` and ``input2`` are multi-dimensional Fortran arrays, with dimension/shape/datatype information automatically inferred in the interface. The ``torchfort_tensor_list_add_tensor`` function adds tensor data by reference
+so changes to the externally provided memory buffer will modify the tensors stored in the list. This is convenient as it enables creating these tensor lists once and reusing them as your program updates the underlying values.
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply docs/usage.rst:164]: This is definitely a bit tricky. One of the perks of TorchFort is that we do not copy input/label data in the existing train/inference API and I'd like to preserve that behavior for the tensor lists as well. I sort of view the tensor list as just a shallow wrapper around data references that ultimately the user owns (and wrote the code to act that way). I'm not sure if having two modes would just make this more confusing.

@nloppi Do you have any thoughts on this?

[the remark this answers is not in the document]
Comment (azrael417) [reply docs/usage.rst:164]: Yeah, I agree it could confuse the user. It would be good to get @nloppi opinion on this as well.
